pkg/patterns/declarative/reconciler.go

```diff
@@ -64,28 +72,69 @@ type DeclarativeObject interface {
 var kubectl = kubectlcmd.New()
 
 func (r *Reconciler) Init(mgr manager.Manager, prototype DeclarativeObject, opts ...reconcilerOption) error {
+	log := log.Log
+
 	r.prototype = prototype
 	r.kubectl = kubectl
 
 	r.client = mgr.GetClient()
 	r.config = mgr.GetConfig()
 	r.mgr = mgr
+	globalObjectTracker.mgr = mgr
 
 	if err := r.applyOptions(opts...); err != nil {
 		return err
 	}
 
-	return r.validateOptions()
+	if err := r.validateOptions(); err != nil {
+		return err
+	}
+
+	if r.CollectMetrics() {
+		if gvk, err := apiutil.GVKForObject(prototype, r.mgr.GetScheme()); err != nil {
+			return err
+		} else {
+			r.gvk = gvk
+			if err := r.tryToSetCounter(); stdliberr.Is(err, &meta.NoKindMatchError{}) {
+				log.WithName("declarative_reconciler").Info("failed to get corresponding RESTMapper from API server. try later")
+			} else if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
 }
 
 // +rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
-func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+func (r *Reconciler) Reconcile(request reconcile.Request) (result reconcile.Result, err error) {
 	ctx := context.TODO()
 	log := log.Log
 
+	if r.CollectMetrics() {
+		if r.setCounter {
+			r.rc.reconcileWith(request)
+			defer func() {
+				if err != nil {
+					r.rfc.failureWith(request)
+				}
+			}()
+		} else if err = r.tryToSetCounter(); err == nil {
+			r.rc.reconcileWith(request)
+			defer func() {
+				if err != nil {
+					r.rfc.failureWith(request)
+				}
+			}()
+		} else {
+			log.WithName("declarative_reconciler").Error(err, "failed to get corresponding RESTMapper from API server. try next time")
```

Comment:
Nit: I'm not sure what "try next time" is implying ... does it mean "will retry" ?

Reply:
Yes. This means "will try". I'll modify it to "Reconciler will try to get RESTMapper next time."